### ISSUE TITLE
How to debug InvalidReqRange error and tikv-ctl usage unclear?

### ISSUE THREAD

**Issue by georgeteo** [tikv/tikv#8626]
## Question

Context: 

- Using the rust client
- Setup TiKV on laptop via `tiup playground`

When scanning the keyspace via rust client:
```
        let mut transaction = self.db.begin().await?;
        let from: Key = b"".to_vec().into();
        let data = transaction.scan(from.., 100, false).await?;
        data.map(|kv| TopicMetadata::decode(Bytes::copy_from_slice(&kv.into_value())))
            .filter_map(Result::ok)
            .collect()
```

It fails with the following error:
```
KeyError { locked: None, retryable: "", abort: "Txn(InvalidReqRange { start: Some([0, 0, 0, 0, 0, 0, 0, 0, 247]), end: Some([116, 128, 0, 0, 0, 0, 0, 0, 255, 255, 5, 0, 0, 0, 0, 0, 0, 255, 0, 248, 0, 0, 0, 0, 0, 0, 249]), lower_bound: Some([]), upper_bound: Some([116, 128, 0, 0, 0, 0, 0, 0, 255, 5, 0, 0, 0, 0, 0, 0, 0, 248]) })", conflict: None, already_exist: None, deadlock: None, commit_ts_expired: None, txn_not_found: None }
```

In an effort to debug this, I'm trying to scan tikv via `tikv-ctl` with the following command:
```
tikv-ctl --pd localhost:2379 raw-scan --from 'z'
TiKV Control (tikv-ctl)
Release Version:   4.0.5
Edition:           Community
Git Commit Hash:   f39927a3529d40a6bb4e6c54854a94fdac996e92
Git Commit Branch: HEAD
UTC Build Time:    2020-09-03 04:54:51
Rust Version:      rustc 1.42.0-nightly (0de96d37f 2019-12-19)
Enable Features:   jemalloc sse protobuf-codec
Profile:           debug
A tool for interacting with TiKV deployments.
USAGE:
    TiKV Control (tikv-ctl) [FLAGS] [OPTIONS] [SUBCOMMAND]
FLAGS:
    -h, --help                    Prints help information
        --skip-paranoid-checks    Skip paranoid checks when open rocksdb
    -V, --version                 Prints version information
OPTIONS:
        --ca-path <ca_path>              Set the CA certificate path
```

There are several questions for tikv:
- Based on `tikv-ctl` error message, I think I'm using the `raw-scan` command correctly, but it doesn't seem to be working.
- The `tikv-ctl` error message could be improved to explain what exactly is wrong with the command rather than printing the whole error message.

In summary, the high level questions are:
- What does `Txn(InvalidReqRange...` mean?
- How do I correctly scan from a specific key in TiKV rust client?
- What error is thrown if the key doesn't exist? Does it "round up" to the next key.
- How do I do a full scan of the TiKV key space?
- What is the correct command for `tikv-ctl`?

I've asked for help on [slack](https://tikv-wg.slack.com/archives/CDDBKTQ9X/p1599497565038500) but haven't gotten much visibility.

**Comment by MyonKeminta** [comment]
I think this is caused by the rust-client didn't split range into regions correctly.
@sticnarf  PTAL thanks

**Comment by georgeteo** [comment]
@MyonKeminta: thanks for the quick response! Any suggestion for scanning the keyspace via `tikv-ctl` for adhoc debugging?

**Comment by sticnarf** [comment]
It looks like wrong handling in raw/encoded key.
And there is also a TODO for transactional scan. https://github.com/tikv/client-rust/blob/d7bded1fbf341f34cdd2b76e4a7206c28e89d484/src/request.rs#L194 So the result set can be larger than expected.
@ekexium Could you take a look at it?

**Comment by MyonKeminta** [comment]
@georgeteo It seems that scanning keys doesn't help much in your problem. If you want, you can try this: https://docs.pingcap.com/tidb/stable/tikv-control#scan-to-view-mvcc-of-a-specific-range
But you need to know a little about how TiKV stores keys internally to understand tikv-ctl's output. The key shown by tikv-ctl is [memory-comparable encoded](https://github.com/facebook/mysql-5.6/wiki/MyRocks-record-format#memcomparable-format) and then prefixed with a letter `'z'`.

**Comment by ekexium** [comment]
> What does Txn(InvalidReqRange... mean?

It means the request range exceeds the bound.

> How do I correctly scan from a specific key in TiKV rust client?

Confirmed there is a bug in client-rust concerning ranges. We plan to fix this recently.

> What error is thrown if the key doesn't exist? Does it "round up" to the next key.

`scan` does not require the existence of the key. It returns all pairs lying in the specified range.

> How do I do a full scan of the TiKV key space?

As far as I know, you can set the upper bound large enough, depending on your scenario. I don't know if there exists a better way.

> What is the correct command for tikv-ctl?

SIG-tools could help improve this. You can also file an issue in the tiup repo.

**Comment by georgeteo** [comment]
> I think this is caused by the rust-client didn't split range into regions correctly.
> @sticnarf PTAL thanks

Does this mean the client is responsible for splitting a scan into multiple parallel requests to each region/node? And that this is not implemented correctly in rust-client?

**Comment by ekexium** [comment]
> Does this mean the client is responsible for splitting a scan into multiple parallel requests to each region/node? And that this is not implemented correctly in rust-client?

Yes

**Comment by georgeteo** [comment]
> Yes. There is a little imperfection in mapping ranges to regions, but it shouldn't matter. The bug is caused by an encoding problem.

Is this a quick fix? I'm happy to submit a PR myself with some pointers.

**Comment by ekexium** [comment]
> Is this a quick fix? I'm happy to submit a PR myself with some pointers.

It would be great if you want to help! Otherwise I'm gonna fix this recently.

There are two representations of a Key in TiKV: encoded and raw. 
https://github.com/tikv/tikv/blob/5a9ae6b7fc48564819d259eb2bc1fff3b99ac023/components/txn_types/src/types.rs#L30-L33

It seems to me that client-rust receives the region from pd with encoded keys. But they are incorrectly treated as raw keys by the rust-client. I think we should add appropriate encoding/decoding stuff to the client. A reference is the [client-go](
https://github.com/tikv/client-go/blob/95774393107b33a79e97c7acd231ff26ed7deede/locate/codec.go#L32):

**Comment by georgeteo** [comment]
> > Is this a quick fix? I'm happy to submit a PR myself with some pointers.
> 
> It would be great if you want to help! Otherwise I'm gonna fix this recently.
> 
> There are two representations of a Key in TiKV: encoded and raw.
> https://github.com/tikv/tikv/blob/5a9ae6b7fc48564819d259eb2bc1fff3b99ac023/components/txn_types/src/types.rs#L30-L33
> 
> It seems to me that client-rust receives the region from pd with encoded keys. But they are incorrectly treated as raw keys by the rust-client. I think we should add appropriate encoding/decoding stuff to the client. A reference is the [client-go](https://github.com/tikv/client-go/blob/95774393107b33a79e97c7acd231ff26ed7deede/locate/codec.go#L32):

Sounds reasonably simple. Feel free to fix this before that if you feel like it - otherwise I'll take a look tomorrow (USA time).

**Comment by georgeteo** [comment]
Some quick questions: 

- The [Go implementation](https://github.com/tikv/client-go/blob/95774393107b33a79e97c7acd231ff26ed7deede/locate/codec.go#L26) wraps and delegates to a wrapped [pd.Client](https://github.com/tikv/pd/blob/master/client/client.go#L41). Should we adopt a similar pattern for rust-client or is it safe to keep it simpler and assume that pd client will always want to encode/decode the bytes?
- Any tips for rust protobuf setup in intellij-rust? It's hard to figure out what the generated code for protobuf accessors/setters/constructors are.
![Screen Shot 2020-09-10 at 10 03 45 PM](https://user-images.githubusercontent.com/12107160/92864572-1fb67180-f3b2-11ea-9e99-e5e8c2141211.png)
- It seems [tikv itself has a pd client](https://github.com/tikv/tikv/blob/3f94eb8d1e8f254d184eb2070c4a6f4e98c6b675/components/pd_client/src/client.rs#L151) that doesn't wrap encoding/decoding. Why does it encode/decode and should we share the implementation or duplicate in the rust client?

**Comment by ekexium** [comment]
@georgeteo 
- Transactional clients need encoding/decoding, while raw clients don't. So I prefer wrapping it.
- It's not supported well by rust IDE/plugins. You can use the [tikv doc](https://tikv.github.io/doc/kvproto/index.html).
- I suggest we implement encoding/decoding in rust-client. Codec in TiKV is complicated but can be a reference. 
- TiKV does not need to encode/decode in its pd client. 
  - Keys in raw requests are not encoded. https://github.com/tikv/tikv/blob/dbb2174c3382fcb3b254b56c00765b264e425a57/src/server/service/kv.rs#L1234
  - Keys in transactional requests are encoded in the beginning of processing. https://github.com/tikv/tikv/blob/dbb2174c3382fcb3b254b56c00765b264e425a57/src/server/service/kv.rs#L1124
  - That's why we should not mix raw and transactional clients. Inner TiKV logic and PD does not have to know anything about encoding.

**Comment by georgeteo** [comment]
I have a preliminary implementation in https://github.com/tikv/client-rust/pull/162. I still have to fix some formatting issues and write tests, but please take a look at the high level idea. 

Caveat: I'm relatively new to rust so feel free to comment on Rust patterns.

Also, it seems to pass `cargo test` locally but consistently fails 3/4 Jenkins CI tests. How do I replicate what is running on Jenkins so that I iterate faster on the issues.
